extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/datatransferproject/transfer/microsoft/photos/MicrosoftPhotosImporter.java

```diff
@@ -142,143 +149,233 @@ private String createOneDriveFolder(PhotoAlbum album) throws IOException {
+      chunksRead++;
+    }
+    final int _debugTotalFileSize = totalFileSize;
+    final int _debugChunksRead = chunksRead;
+    // finish timing
+    Instant finish = Instant.now();
+    final long timeElapsed = Duration.between(start, finish).toMillis();
+    monitor.info(()-> String.format("Took %d millseconds to read %s bytes, split into %d chunks", timeElapsed, _debugTotalFileSize, _debugChunksRead));
+
+
+
+    String photoUploadUrl = createUploadSession(photo, idempotentImportExecutor);
+    // PUT to {uploadurl}
+    // HEADERS
+    // Content-Length: {chunk size in bytes}
+    // Content-Range: bytes {begin}-{end}/{total size}
+    // body={bytes}
+    Request.Builder uploadRequestBuilder = new Request.Builder().url(photoUploadUrl);
+    uploadRequestBuilder.header("Authorization", "Bearer " + credential.getAccessToken());
+
+    Integer chunkCode = null;
+    Response chunkResponse = null;
+    for (Chunk chunk : chunksToSend) {
+      // put chunk data in
+      RequestBody uploadChunkBody = RequestBody.create(MediaType.parse(photo.getMediaType()), chunk.getData(), 0, chunk.getSize());
+      uploadRequestBuilder.put(uploadChunkBody);
+      // set chunk data headers, indicating size and chunk range
+      uploadRequestBuilder.header("Content-Length", chunk.getSize().toString());
+      final String contentRange = String.format("bytes %d-%d/%d", chunk.getStart(), chunk.getEnd(), totalFileSize);
+      uploadRequestBuilder.header("Content-Range", contentRange);
+      // upload the chunk
+      chunkResponse = client.newCall(uploadRequestBuilder.build()).execute();
+      // check for auth error
+      chunkCode = chunkResponse.code();
+      if (chunkCode == 401) {
+        // If there was an unauthorized error, then try refreshing the creds
+        credentialFactory.refreshCredential(credential);
+        monitor.info(() -> "Refreshed authorization token successfuly");
+        // update auth info, reupload chunk
+        uploadRequestBuilder.header("Authorization", "Bearer " + credential.getAccessToken());
+        chunkResponse = client.newCall(uploadRequestBuilder.build()).execute();
+        chunkCode = chunkResponse.code();
+      }
+      if (chunkCode < 200 || chunkCode > 299) {
+        throw new IOException(
+          "Got error code: " + chunkCode + " message: " + chunkResponse.message() + " body: " + chunkResponse
+          .body().string());
+      }
+      if (chunkCode == 200) {
+        monitor.info(() -> String.format("Uploaded chunk %s-%s successfuly", chunk.getStart(), chunk.getEnd()));
       }
     }
+    // get complete file response
+    Preconditions.checkState(chunkCode == 201 || chunkCode == 200, "Got bad response code when finishing uploadSession: %d", chunkCode);
+    //
+    ResponseBody chunkResponseBody = chunkResponse.body();
+    Map<String, Object> chunkResponseData = objectMapper.readValue(chunkResponseBody.bytes(), Map.class);
+    return (String) chunkResponseData.get("id");
```

Comment:
For readability purposes can you generalize the concrete blocks of this function so that it's not 100 lines long?

e.g. `List<Chunks> = getpages(photo)`, ` for chunk in chunks: upload_file(chunk)`, that sorta stuff.